File: it-agents/skills/code-architect/references/code-examples/typescript/Observer.ts
// Observer — subscription mechanism to notify subscribers of events.

interface Observer<T> {
  update(payload: T): void;
}

class Subject<T> {
  private observers = new Set<Observer<T>>();

  subscribe(observer: Observer<T>): () => void {
    this.observers.add(observer);
    return () => this.observers.delete(observer); // unsubscribe handle
  }

  notify(payload: T): void {
    for (const observer of this.observers) {
      observer.update(payload);
    }
  }
}

class LoggingObserver implements Observer<number> {
  constructor(private readonly label: string) {}
  update(value: number): void {
    console.log(`${this.label} received: ${value}`);
  }
}

function demo(): void {
  const temperature = new Subject<number>();
  const display = new LoggingObserver("display");
  const unsubscribe = temperature.subscribe(display);
  temperature.subscribe(new LoggingObserver("logger"));

  temperature.notify(21);
  unsubscribe();
  temperature.notify(23); // display no longer notified
}

demo();

File: it-agents/skills/code-architect/references/code-examples/csharp/Observer.cs
// Observer — subscription mechanism to notify subscribers of events.
using System;
using System.Collections.Generic;

namespace Patterns.Observer
{
    interface IObserver
    {
        void Update(int state);
    }

    interface ISubject
    {
        void Subscribe(IObserver observer);
        void Unsubscribe(IObserver observer);
        void Notify();
    }

    sealed class Subject : ISubject
    {
        private readonly List<IObserver> _observers = new();
        private int _state;

        public int State
        {
            get => _state;
            set { _state = value; Notify(); }
        }

        public void Subscribe(IObserver observer) => _observers.Add(observer);
        public void Unsubscribe(IObserver observer) => _observers.Remove(observer);
        public void Notify()
        {
            foreach (IObserver o in _observers.ToArray())
                o.Update(_state);
        }
    }

    sealed class ConcreteObserver : IObserver
    {
        private readonly string _name;
        public ConcreteObserver(string name) => _name = name;
        public void Update(int state) => Console.WriteLine($"{_name} notified of state {state}");
    }

    static class Demo
    {
        public static void Run()
        {
            var subject = new Subject();
            var a = new ConcreteObserver("A");
            var b = new ConcreteObserver("B");
            subject.Subscribe(a);
            subject.Subscribe(b);

            subject.State = 1;
            subject.Unsubscribe(a);
            subject.State = 2;
        }

        public static void Main() => Run();
    }
}
